target/linux/ath79/files/drivers/mfd/rb4xx-cpld.c

```diff
@@ -0,0 +1,208 @@
+ *
+ * The CPLD also acts as a GPIO expander.
+ *
+ * Copyright (C) 2008-2011 Gabor Juhos <juhosg@openwrt.org>
+ * Copyright (C) 2008 Imre Kaloz <kaloz@openwrt.org>
+ * Copyright (C) 2015 Bert Vermeulen <bert@biot.com>
+ * Copyright (C) 2020 Christopher Hill <ch6574@gmail.com>
+ *
+ * This file was based on the driver for Linux 2.6.22 published by
+ * MikroTik for their RouterBoard 4xx series devices.
+*/
+#include <linux/mfd/core.h>
+#include <linux/spi/spi.h>
+#include <linux/module.h>
+#include <linux/of_platform.h>
+
+#include <mfd/rb4xx-cpld.h>
+
+/* CPLD commands */
+#define CPLD_CMD_WRITE_NAND	0x08 /* send cmd, n x send data, send idle */
+#define CPLD_CMD_WRITE_CFG	0x09 /* send cmd, n x send cfg */
+#define CPLD_CMD_READ_NAND	0x0a /* send cmd, send idle, n x read data */
+#define CPLD_CMD_READ_FAST	0x0b /* send cmd, 4 x idle, n x read data */
+#define CPLD_CMD_GPIO8_HIGH	0x0c /* send cmd */
+#define CPLD_CMD_GPIO8_LOW	0x0d /* send cmd */
+
+static int rb4xx_cpld_write_nand(struct rb4xx_cpld *cpld, const void *tx_buf,
+				 unsigned int len)
+{
+	struct spi_message m;
+	static const u8 cmd = CPLD_CMD_WRITE_NAND;
+	struct spi_transfer t[3] = {
+		{
+			.tx_buf = &cmd,
+			.len = sizeof(cmd),
+		}, {
+			.tx_buf = tx_buf,
+			.len = len,
+			.tx_nbits = SPI_NBITS_DUAL,
+		}, {
+			.len = 1,
+			.tx_nbits = SPI_NBITS_DUAL,
+		},
+	};
+
+	spi_message_init(&m);
+	spi_message_add_tail(&t[0], &m);
+	spi_message_add_tail(&t[1], &m);
+	spi_message_add_tail(&t[2], &m);
+	return spi_sync(cpld->spi, &m);
+}
+
+static int rb4xx_cpld_read_nand(struct rb4xx_cpld *cpld, void *rx_buf,
+				unsigned int len)
+{
+	struct spi_message m;
+	static const u8 cmd[2] = {
+		CPLD_CMD_READ_NAND, 0
+	};
+	struct spi_transfer t[2] = {
+		{
+			.tx_buf = &cmd,
+			.len = sizeof(cmd),
+		}, {
+			.rx_buf = rx_buf,
+			.len = len,
+		},
+	};
+
+	spi_message_init(&m);
+	spi_message_add_tail(&t[0], &m);
+	spi_message_add_tail(&t[1], &m);
+	return spi_sync(cpld->spi, &m);
+}
+
+static int rb4xx_cpld_cmd(struct rb4xx_cpld *cpld, const void *tx_buf,
+			  size_t len)
+{
+	struct spi_message m;
+	struct spi_transfer t = {
+		.tx_buf = tx_buf,
+		.len = len,
+	};
+
+	spi_message_init(&m);
+	spi_message_add_tail(&t, &m);
+	return spi_sync(cpld->spi, &m);
+}
+
+static int rb4xx_cpld_gpio_set(struct rb4xx_cpld *cpld, unsigned int offset,
+			       int value)
+{
+	u16 values = cpld->values;
+	u8 lsb, msb, cmd[2];
+	int ret;
+
+	if (value)
+		values |= BIT(offset);
+	else
+		values &= ~(BIT(offset));
+
+	lsb = values & 0xff; /* GPIO 0-7 */
+	msb = values >> 8;   /* GPIO 8   */
+
+	if (lsb != (cpld->values & 0xff)) {
+		/* GPIO 0-7 change can be sent direct to CPLD */
+		cmd[0] = CPLD_CMD_WRITE_CFG;
+		cmd[1] = lsb;
+
+		ret = rb4xx_cpld_cmd(cpld, &cmd, 2);
+		if (unlikely(ret))
+			return ret;
+	} else if (msb != (cpld->values >> 8)) {
+		/* GPIO 8 is a special case */
+		cmd[0] = CPLD_CMD_GPIO8_HIGH | !!(msb);
```

Comment:
Correct, we just send the magic led5 high/low command out by itself.

Reply:
What a shitty hardware implementation :)
I'll think about this some more later

Reply:
N.B. The if/else if logic could be driven by `offset` instead of comparing changes in the bitfield if that's easier to read.

As in

```
if (offset <8)
  // send WRITE_CFG on SPI
else if (offset == 8)
  // send magic GPIO8_HIGH
```

Reply:
And thinking some more, I wouldn't need to keep state of the 8th gpio, as that would always be passed in, meaning the `u16 values` could shrink to a u8, and I could remove some of the shifting.

Thinking aloud, I now like that better.